src/main/java/uk/ac/ebi/ampt2d/accession/service/MonotonicAccessionGenerator.java

```diff
@@ -0,0 +1,241 @@
+     */
+    private synchronized void increaseAvailableRangesUntilSizeIs(int totalAccessionsToGenerate) {
+        while (availableRanges.getTotalOfValuesInQueue() < totalAccessionsToGenerate) {
+            try {
+                ExponentialBackOff.execute(() -> reserveNewBlock(categoryId, instanceId, blockSize), 10, 30);
+            } catch (RuntimeException e) {
+                // Ignore, max backoff have been reached, we will try again until we can reserve blocks
+            }
+        }
+    }
+
+    /**
+     * Initialize the block service loading previously reserved blocks
+     */
+    @Override
+    public void afterPropertiesSet() {
+        List<ContiguousIdBlock> uncompletedBlocks = getUncompletedBlocksForThisInstanceOrdered();
+        //Insert as available ranges
+        for (ContiguousIdBlock block : uncompletedBlocks) {
+            addBlock(block);
+        }
+    }
+
+    /**
+     * TODO this can be changed to a query to recover and avoid the filter
+     */
+    private List<ContiguousIdBlock> getUncompletedBlocksForThisInstanceOrdered() {
+        try (Stream<ContiguousIdBlock> reservedBlocksOfThisInstance = contiguousIdBlockRepository
+                .findAllByCategoryIdAndInstanceIdOrderByEndAsc(categoryId, instanceId)) {
+            return reservedBlocksOfThisInstance.filter(ContiguousIdBlock::isNotFull).collect(Collectors.toList());
+        }
+    }
+
+    @Transactional(isolation = Isolation.SERIALIZABLE)
+    protected synchronized void reserveNewBlock(String categoryId, String instanceId, long size) {
+        ContiguousIdBlock lastBlock = contiguousIdBlockRepository.findFirstByCategoryIdOrderByEndDesc(categoryId);
+        if (lastBlock == null) {
+            addBlock(contiguousIdBlockRepository.save(new ContiguousIdBlock(categoryId, instanceId, 0, size)));
+        } else {
+            addBlock(contiguousIdBlockRepository.save(lastBlock.nextBlock(instanceId, size)));
+        }
+    }
+
+    private void addBlock(ContiguousIdBlock block) {
+        activeBlocks.offer(block);
+        availableRanges.add(new MonotonicRange(block.getLastCommitted() + 1, block.getEnd()));
+    }
+
+    public MonotonicRangePriorityQueue getAvailableRanges() {
+        return availableRanges;
+    }
+
+    public synchronized void recoverState(long[] committedElements) throws AccessionIsNotPending {
+        List<MonotonicRange> ranges = MonotonicRange.convertToMonotonicRanges(committedElements);
+        List<MonotonicRange> newAvailableRanges = new ArrayList<>();
+        for (MonotonicRange monotonicRange : this.availableRanges) {
+            newAvailableRanges.addAll(monotonicRange.excludeIntersections(ranges));
+        }
+
+        this.availableRanges.clear();
+        this.availableRanges.addAll(newAvailableRanges);
+        doCommit(committedElements);
+    }
+
+    public synchronized void commit(long... accessions) throws AccessionIsNotPending {
+        assertAccessionIsPending(accessions);
+
```

Comment:
Remove blank line, 2 lines put together are easier to read than with it in the middle.

Reply:
No longer applies